### ISSUE TITLE
Alt+F2 spawns new wsltty window as a separate window stash in the task bar

### ISSUE THREAD

**Issue by ksze** [mintty/wsltty#96]
wsltty 1.8.4
Windows 10 Professional 64-bit version 1709, build 16299.248

Just like the title says.

I use the unmodified "WSL Terminal" shortcut (without "%"), which launches bash of my default Linux flavour (Ubuntu).

Once the first wsltty window has been opened, if I press the Alt+F2 combo, I get a new window. But this second window and any subsequent ones become a separate window stash in the task bar:

![image](https://user-images.githubusercontent.com/381556/37010825-5364f1be-2128-11e8-9b24-88a27ef04688.png)

It's slightly annoying. Sometimes I check the window previews in the task bar and don't find one window that I expect in the stash (because the first window is not in the stash), and it confuses me for a moment.

If I check the process tree in Process Explorer (from the SysInternals Suite), I see that the first mintty has explorer.exe as parent, while the second and subsequent mintty processes have "no" parent:

![image](https://user-images.githubusercontent.com/381556/37010753-f035409e-2127-11e8-86d7-18ff85330cba.png)

**Comment by mintty** [comment]
There were some bothersome tweaks about taskbar behaviour in mintty quite a while ago (before wsltty) and I hoped this trouble would not occur again...
It will take a while until I find some time to check this out.

**Comment by mintty** [comment]
I wonder why this only happens in WSL mode. No idea how to prevent it.

**Comment by willpower232** [comment]
I use a custom shortcut where the only difference is that I start in a different directory to avoid near constant `cd`'s. The shortcut is copied from the ubuntu one so it has the same icon.

When I press Alt+F2 I get a new stash of windows too, in the same custom directory, without the ubuntu icon so it appears as the tux icon instead (as expected from above).

However when I middle mouse button click the ubuntu logo, I get another window in the same stash with the same icon and path so apparently theres a difference between middle mouse button and Alt+F2. I'd assume the only reason it isn't grouped is the icon.

HTH

**Comment by mintty** [comment]
Workaround: add `-o AppID=wsltty` to the target command line in the shortcut, as a mintty option (i.e. before wslbridge). Any AppID will do.

**Comment by mintty** [comment]
> The shortcut is copied from the ubuntu one

The shortcut may have its own "AppID" which is then inherited by mintty. Please try another shortcut, maybe a freshly-created one.

**Comment by willpower232** [comment]
>Workaround: add -o AppID=wsltty to the target command line in the shortcut, as a mintty option (i.e. before wslbridge). Any AppID will do.

That seems to solve the keyboard issue but also stops the middle mouse button from opening a new window. FWIW I can't see any AppID's on the existing shortcuts.

**Comment by willpower232** [comment]
I referred to the original issue as the keyboard issue because the same feature of opening a new window can also happen on middle mouse click.

Without an AppID, I can open a new window in the same stack no problems using the middle mouse button.

With an AppID, I can't open a new window at all using the middle mouse click but Alt+F2 works fine as you described.

**Comment by mintty** [comment]
Released 1.9.0.

**Comment by sagarrabadiya** [comment]
i am using 1.9.0 still it shows multiple windows in task bar once i pin the shortcut to task bar clicking on it, will always create new icon for ex.
![image](https://user-images.githubusercontent.com/5405403/42210197-2ad15e14-7ece-11e8-8dd2-4687a7680e98.png)

**Comment by mintty** [comment]
The original issue was about Alt+F2, does that work now for you?
Taskbar pinning is another issue, as Microsoft designed a completely separate, insane concept for its configuration.

**Comment by sagarrabadiya** [comment]
yes with Alt + F2 it doesn't show multiple icon, so i think problem is with pin to task bar, can we have any work around or solution for that?

**Comment by mintty** [comment]
You may try to follow the hints given in the manual page for options `TaskCommands` and `AppID`.
This crap only works if you make it irreversibly persistent...

**Comment by mintty** [comment]
Released wsltty 1.9.2 with new option AppID=@.

**Comment by mintty** [comment]
Planning to improve the situation for the next release in two ways:
* Alt+F2 in a window started from a shortcut could invoke the shortcut rather than just the binary (.exe), thus inheriting the same AppID from the shortcut as the running instance.
* Add placeholders to option AppID to be able to use it but distinguish various deployments, invocations etc as preferred (somewhat covered by value AppID=@ already but more flexible).

**Comment by ksze** [comment]
Wow. The lengths you go to to fix the quirks.

**Comment by mintty** [comment]
I've uploaded a patch that derives the AppID from the desktop shortcut if mintty is started from a shortcut (desktop, taskbar or start menu).
Currently it takes precedence over an explicitly defined AppID option.
Please test if possible (building mintty in cygwin and replacing the binary in %LOCALAPPDATA%/wsltty/bin).

**Comment by mintty** [comment]
Taskbar icon configuration is one of the extremely obscure areas of Windows, handled by insanely obfuscated APIs. I'm not surprised about erratic behaviour here.
